OSSM-8986: Cluster-wide migration using istio injection label with ce… (#89488)
openshift/openshift-docs: modules/ossm-migrating-a-cluster-wide-deployment-using-the-istio-injection-label-with-cert-manager.adoc

* You have the {SMProduct} {SMv2Version} Operator installed.
* You have the {SMProduct} 3 Operator installed.
* You created an `IstioCNI` resource.
* You have the `istioctl` tool installed.
* You are running a `ClusterWide` `ServiceMeshControlPlane`.
* You have installed the `bookinfo` application.

.Procedure

. Check that your {SMProduct} 2 `ServiceMeshControlPlane` is configured with the cert-manager-tool:
+
.Example `ServiceMeshControlPlane` cert-manager configuration
[source,yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  ...
  security:
    certificateAuthority:
      cert-manager:
        address: cert-manager-istio-csr.istio-system.svc:443
      type: cert-manager
    dataPlane:
      mtls: true
    identity:
      type: ThirdParty
    manageNetworkPolicy: false
----

. Update the `istio-csr` deployment to include your {SMProduct} 3 control plane by running the following command:
+
[source,terminal]
----
  helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
      --install \
      --reuse-values \
      --namespace istio-system \
      --wait \

>       --set "app.istio.revisions={basic,istio-tenant-a}" <1>

Author: Done

----
+
<1> The `app.istio.revisions` field needs to include your {SMProduct} 3.0 control plane revision _before_ you create your `Istio` resource so that proxies can properly communicate with the {SMProduct} 3.0 control plane.

. Create a YAML file named `ossm-3.yaml` that creates the {istio} resource for the 3.0 installation in the same namespace as the `ServiceMeshControlPlane` resource for the 2.6 installation.
+
.Example `Istio` resource
[source,yaml,subs="attributes,verbatim"]
----
apiVersion: sailoperator.io/v1alpha1
kind: Istio
metadata:
  name: ossm-3 <1>
spec:
  updateStrategy:
    type: RevisionBased
  namespace: istio-system <2>
  version: v1.24.3
  values:  
    meshConfig:
    extensionProviders: <3>
      - name: prometheus
        prometheus: {}
      - name: otel
        opentelemetry:
          port: 4317
          service: otel-collector.opentelemetrycollector-3.svc.cluster.local
  global:
      caAddress: cert-manager-istio-csr.istio-system.svc:443
      pilot:
        env:

>           ENABLE_CA_SERVER: "false"

Author: Done

<3> Optional: If you are migrating metrics and tracing, update the `extensionProviders` fields according to your tracing and metrics configurations.
+
This example configuration does not use any `discoverySelectors`, which means the {istio} control plane has access to all namespaces. If you want to define `discoverySelectors` all data plane namespaces that you plan to migrate from 2.6 must be matched.
+
[NOTE]
====
Namespaces that have `istio-injection=enabled` label applied and are managed by {SMProduct} 2.6 control plane. To prevent the {SMProduct} 3.0 control pane from injecting proxies in workloads, do not use use the `default` name for the {istio} resource, and do not create the `default` revision tag at this time. You will create the `default` revision tag later in this procedure.
====

. Identify the namespaces that contain a 2.6 control plane by running the following command:
+
[source,terminal]
----
$ oc get smcp -A
----
+
.Sample output:
[source,terminal]
----
NAMESPACE      NAME                   READY   STATUS            PROFILES      VERSION   AGE
istio-system   install-istio-system   6/6     ComponentsReady   ["default"]   2.6.6     115m

> ----

Reviewer: I think the order of things got jumbled a bit. This step needs to come after updating istio-csr and before "Create a YAML file named `ossm-3.yaml`" since users will use the `namespace` output from this step in creating their `Istio` yaml.

Author: Moved this step.

// Module included in the following assemblies:
//
// * service-mesh-docs-main/migrating/cluster-wide/ossm-migrating-cluster-wide-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-migrating-a-cluster-wide-deployment-using-the-istio-injection-label-with-cert-manager_{context}"]
= Migrating a cluster-wide deployment by using the istio injection label with cert-manager

You can perform a canary upgrade with the gradual migration of data plane namespaces for a cluster-wide deployment by using the `istio-injection=enabled` label and the `default` revision tag. 

> You must re-label all of the data plane namespaces. However, it is safe to restart any of the workloads at any point during the migration process. 

Reviewer 2: s/re-label/relabel??

Author: done

----

. Create a YAML file named `ossm-3.yaml`. This procedure creates the {istio} resource for the 3.0 installation in the same namespace as the `ServiceMeshControlPlane` resource for the 2.6 installation.
+
[NOTE]
====
In the following example configuration, the {istio} control plane has access to all namespaces on the cluster. If you want to limit the namespaces the control plan has access to, you must define discovery selectors. You must match all the data plane namespaces that you plan to migrate from version 2.6.
====
+
.Example `Istio` resource
[source,yaml,subs="attributes,verbatim"]
----
apiVersion: sailoperator.io/v1
kind: Istio
metadata:
  name: ossm-3 # <1>
spec:
  updateStrategy:
    type: RevisionBased
  namespace: istio-system # <2>
  version: v1.24.3
  values:  
    meshConfig:
      extensionProviders: # <3>
        - name: prometheus
          prometheus: {}
        - name: otel
          opentelemetry:
            port: 4317
            service: otel-collector.opentelemetrycollector-3.svc.cluster.local
    global:
      caAddress: cert-manager-istio-csr.istio-system.svc:443

>       pilot:
>         env:
>           ENABLE_CA_SERVER: "false"

Reviewer 3: ```suggestion
    pilot:
      env:
        ENABLE_CA_SERVER: "false"
```

Author: done

validated using https://yamlchecker.com/

----

> <1> The `name`, `updateStrategy` and `version` fields specify how the `IstioRevision` resource name is created. For more information, see "Identifying the revision name".

Reviewer 2: ```suggestion
<1> The `name`, `updateStrategy`, and `version` fields specify how the `IstioRevision` resource name is created. For more information, see "Identifying the revision name".
```

Author: done

<2> The 3.0 and 2.6 control planes must run in the same namespace.
<3> If you are migrating metrics and tracing, update the `extensionProviders` fields according to your tracing and metrics configurations.
+
[NOTE]
====

> Namespaces that have `istio-injection=enabled` label applied and are managed by {SMProduct} 2.6 control plane. To prevent the {SMProduct} 3.0 control pane from injecting proxies in workloads, do not use use the `default` name for the {istio} resource, and do not create the `default` revision tag at this time. You will create the `default` revision tag later in this procedure.

Reviewer 3: ```suggestion
To prevent the OpenShift Service Mesh 3.0 control plane from injecting proxies in the namespaces that have the istio-injection=enabled label applied and are still managed by OpenShift Service Mesh 2.6 control plane, do not use use the default name for the Istio resource, and do not create the default revision tag in the following steps. You create the default revision tag later in this procedure.
```

Author: done

> To prevent the OpenShift Service Mesh 3.0 control plane from injecting proxies in the namespaces that have the istio-injection=enabled label applied and are still managed by OpenShift Service Mesh 2.6 control plane, do not use the default name for the Istio resource, and do not create the default revision tag in the following steps. You create the default revision tag later in this procedure.

Reviewer 2: > and are still managed by OpenShift Service Mesh 2.6 control plane
I'm not sure what this refers to, what is managed by the Service Mesh control plane.

> To prevent the OpenShift Service Mesh 3.0 control plane from injecting proxies in the namespaces that have the istio-injection=enabled label applied and are still managed by OpenShift Service Mesh 2.6 control plane, do not use the default name for the Istio resource, and do not create the default revision tag in the following steps. You create the default revision tag later in this procedure.

Author: It refers to the proxies. the proxies. I rephrased it.

> To prevent the OpenShift Service Mesh 3.0 control plane from injecting proxies in the namespaces that have the istio-injection=enabled label applied and are still managed by OpenShift Service Mesh 2.6 control plane, do not use use the default name for the Istio resource, and do not create the default revision tag in the following steps. You create the default revision tag later in this procedure.

Author: done
